pkg/serializer/jsonstream/compressor.go

```diff
@@ -0,0 +1,258 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+//+build zlib
+
+package jsonstream
+
+import (
+	"bytes"
+	"compress/zlib"
+	"errors"
+	"expvar"
+	"sync"
+
+	"github.com/DataDog/datadog-agent/pkg/forwarder"
+	"github.com/DataDog/datadog-agent/pkg/serializer/marshaler"
+	"github.com/DataDog/datadog-agent/pkg/util/compression"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+const (
+	// Available is true if the code is compiled in
+	Available = true
+)
+
+var (
+	expvars              = expvar.NewMap("jsonstream")
+	expvarsTotalCalls    = expvar.Int{}
+	expvarsTotalItems    = expvar.Int{}
+	expvarsTotalPayloads = expvar.Int{}
+	expvarsTotalCycles   = expvar.Int{}
+	expvarsItemDrops     = expvar.Int{}
+	expvarsBytesIn       = expvar.Int{}
+	expvarsBytesOut      = expvar.Int{}
+)
+
+func init() {
+	expvars.Set("TotalCalls", &expvarsTotalCalls)
+	expvars.Set("TotalItems", &expvarsTotalItems)
+	expvars.Set("TotalPayloads", &expvarsTotalPayloads)
+	expvars.Set("TotalCompressCycles", &expvarsTotalCycles)
+	expvars.Set("ItemDrops", &expvarsItemDrops)
+	expvars.Set("BytesIn", &expvarsBytesIn)
+	expvars.Set("BytesOut", &expvarsBytesOut)
+}
+
+// the backend accepts payloads up to 3MB/50MB, but being conservative is okay
+var (
+	megaByte            = 1024 * 1024
+	maxPayloadSize      = 2*megaByte + megaByte/2
+	maxUncompressedSize = 45 * megaByte
+	maxRepacks          = 40
+)
+
+var (
+	errPayloadFull = errors.New("reached maximum payload size")
+	errTooBig      = errors.New("item alone exceeds maximum payload size")
+)
+
+var jsonSeparator = []byte(",")
+
+// inputBufferPool is an object pool of inputBuffers
+var inputBufferPool = sync.Pool{
+	New: func() interface{} {
+		return bytes.NewBuffer(make([]byte, 0, maxPayloadSize))
```

Comment:
I think it's worth explaining the rationale behind this initial allocation of `maxPayloadSize`

Reply:
Updated the comment